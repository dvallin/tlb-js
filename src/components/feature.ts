import { Color } from '../renderer/color'
import { gray, primary } from '../renderer/palettes'
import { TlbWorld } from '../tlb'

import { FeatureComponent } from './feature'
import { PositionComponent } from './position'
import { WorldMap } from '../resources/world-map'
import { Vector } from '../spatial'
import { Entity } from '../ecs/entity'
import { strangeSymbols } from '../symbols'

export type FeatureType = keyof typeof features
export interface FeatureComponent {
  type: FeatureType
}

export interface Feature {
  character: string
  diffuse: Color

  blocking: boolean
  lightBlocking: boolean
  description: string
}

export const features = {
  wall: {
    character: '#',
    diffuse: gray[3],
    blocking: true,
    lightBlocking: true,
    description: 'a wall',
  },
  corridor: {
    character: '.',
    diffuse: gray[0],
    blocking: false,
    lightBlocking: false,
    description: 'a corridor',
  },
  room: {
    character: '.',
    diffuse: primary[1],
    blocking: false,
    lightBlocking: false,
    description: 'floor of a room',
  },
  locker: {
    character: strangeSymbols[16],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description: 'a locker',
  },
  trash: {
    character: strangeSymbols[21],
    diffuse: gray[2],
    blocking: false,
    lightBlocking: false,
    description: 'some trash',
  },
  door: {
    character: strangeSymbols[27],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description: 'a door',
  },
  hub: {
    character: '.',
    diffuse: primary[0],
    blocking: false,
    lightBlocking: false,
    description: 'floor of a hub',
  },
  player: {
    character: '@',
    diffuse: primary[0],
    blocking: true,
    lightBlocking: true,
    description: 'you',
  },
  guard: enemy('g', 'a guard'),
  eliteGuard: eliteEnemy('g', 'a very strong guard'),
}

function enemy(character: string, description: string): Feature {
  return {
    character,
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description,
  }
}

function eliteEnemy(character: string, description: string): Feature {
  return {
    character,
    diffuse: primary[3],
    blocking: true,
    lightBlocking: true,
    description,
  }
}

export function getFeature(world: TlbWorld, entity: number): Feature | undefined {
  const feature = world.getComponent<FeatureComponent>(entity, 'feature')
  if (feature) {
    return features[feature.type]
  }
  return undefined
}

export function createFeature(world: TlbWorld, map: WorldMap, position: Vector, type: FeatureType): Entity {
  const e = map.getTile(position)
  if (e !== undefined) {
    throw new Error(`there is already a tile at ${position.key}`)
  }

  const entity = world
    .createEntity()
    .withComponent<PositionComponent>('position', { position })
    .withComponent<FeatureComponent>('feature', { type }).entity
  map.setTile(position, entity)
  return entity
}
