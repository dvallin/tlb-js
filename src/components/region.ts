import { Shape } from '../geometry/shape'
import { Vector } from '../spatial'
import { Entity } from '../ecs/entity'

export type RegionsType = 'elevator' | 'red'

export interface RegionComponent {
  type: RegionsType
  level: number
  shape: Shape
  exits: Vector[]
  authorized: Set<Entity>
}

export type StructureType = 'room' | 'hub' | 'corridor'

export interface StructureComponent {
  region: Entity
  kind: StructureType
  shape: Shape
  connections: Entity[]
}
