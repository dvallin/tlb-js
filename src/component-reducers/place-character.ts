import { Entity } from '../ecs/entity'
import { PositionComponent, centeredPosition } from '../components/position'
import { TlbWorld } from '../tlb'
import { WorldMapResource } from '../resources/world-map'
import { Vector } from '../spatial'

export function placeCharacter(world: TlbWorld, entity: Entity, level: number, position: Vector): void {
  const map = world.getResource<WorldMapResource>('map')
  map.levels[level].setCharacter(position, entity)
  world.editEntity(entity).withComponent<PositionComponent>('position', centeredPosition(level, position))
}
