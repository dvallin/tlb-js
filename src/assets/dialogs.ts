import { TlbWorld } from '../tlb'
import { Entity } from '../ecs/entity'
import { TriggeredByComponent, TriggersComponent } from '../components/trigger'
import { DialogComponent } from '../components/dialog'
import { PositionComponent } from '../components/position'
import { WorldMap, WorldMapResource, Level } from '../resources/world-map'
import { coveringDirections } from '../component-reducers/cover'
import { hasItem } from '../component-reducers/inventory'
import { getTriggersOfStructure } from '../component-reducers/structure'
import { LoreComponent } from '../components/lore'
import { quests } from './quests'
import { QuestComponent } from '../components/quest'
import { lores } from './lores'

export type AnswerType = 'close' | 'attack' | 'authorized' | 'move_level_up' | 'move_level_down' | 'trigger' | 'unlock_lore' | 'start_quest'
export type Result = Answer | Navigation | NextDialog
export interface NextDialog {
  text: string
  type: 'next_dialog'
  dialog: DialogType
}
export interface Answer {
  text: string
  type: AnswerType
  target?: Entity
}
export interface Navigation {
  text: string
  type: number
}
export interface Step {
  text: string[]
  answers: Result[]
}
export interface Dialog {
  steps(world: TlbWorld, player: Entity, owner: Entity): Step[]
}

function navigation(text: string, type: number): Navigation {
  return { text, type }
}
function nextDialog(text: string, dialog: DialogType): NextDialog {
  return { text, type: 'next_dialog', dialog }
}
function answer(text: string, type: AnswerType, target?: Entity): Answer {
  return { text, type, target }
}
function attack(): Answer {
  return answer('[attack]', 'attack')
}
function moveUp(): Answer {
  return answer('[move one level up]', 'move_level_up')
}
function moveDown(): Answer {
  return answer('[move one level down]', 'move_level_down')
}

function statelessDialog(steps: Step[]): Dialog {
  return { steps: () => steps }
}

function dialog(steps: (world: TlbWorld, player: Entity, owner: Entity) => Step[]): Dialog {
  return { steps }
}

function plaintextDialog(text: string, type: AnswerType = 'close'): Dialog {
  return statelessDialog([
    {
      text: [text],
      answers: [answer('ok', type)],
    },
  ])
}

function doorControls(world: TlbWorld, player: Entity): Step {
  const doorControlsAnswers = []

  // use player position here instead of owner, because owner might not have a position
  const position = world.getComponent<PositionComponent>(player, 'position')!
  const map: WorldMap = world.getResource<WorldMapResource>('map')
  const level: Level = map.levels[position.level]
  const structure = level.getStructure(position.position)!
  const doors = getTriggersOfStructure(world, structure, ['door'])

  doors.forEach(door => {
    const triggers = world.getComponent<TriggersComponent>(door, 'triggers')!
    const p = world.getComponent<PositionComponent>(triggers.entities[0], 'position')!
    const directions = coveringDirections(p.position, position.position)
    doorControlsAnswers.push(answer(`a door to the ${directions.join(',')}`, 'trigger', door))
  })

  doorControlsAnswers.push(navigation('back', 0))
  return {
    text: ['door controls'],
    answers: doorControlsAnswers,
  }
}

const dialogDefinitions = {
  startLore1: plaintextDialog('here have some lore...', 'unlock_lore'),
  startQuest1: plaintextDialog('please fetch me something...', 'start_quest'),
  terminal: dialog((world, player, owner) => {
    const mainMenuNavigation: (Navigation | NextDialog)[] = []
    const steps = []
    const lore = world.getComponent<LoreComponent>(owner, 'lore')
    if (lore !== undefined) {
      const dialog = lores[lore.type]
      mainMenuNavigation.push(nextDialog(dialog.startText, dialog.start))
    }

    steps.push(doorControls(world, player))
    mainMenuNavigation.push(navigation('door controls', steps.length))

    return [
      {
        text: ['terminal'],
        answers: [...mainMenuNavigation, answer('exit', 'close')],
      },
      ...steps,
    ]
  }),
  civilianDialog: dialog((world, _player, owner) => {
    const mainMenuNavigation: (Navigation | NextDialog)[] = []
    const quest = world.getComponent<QuestComponent>(owner, 'quest')
    if (quest !== undefined) {
      const { startText, start } = quests[quest.type]
      mainMenuNavigation.push(nextDialog(startText, start))
    }

    return [
      {
        text: ['terminal'],
        answers: [...mainMenuNavigation, answer('exit', 'close')],
      },
      {
        text: civilianRandomRemarks,
        answers: [navigation('ok', 0)],
      },
    ]
  }),
  guardRandomRemarks: statelessDialog([
    {
      text: [
        'Do you want to cry now?',
        'Love a neurogender bitch, oh, it get my dick hard',
        "You ain't gonna let me fuck you and I feel you",
        "I'm icy bitch, don't look at my wrist.",
        'Fuck love. All I got for hoes is hard dick and bubblegum.',
        'I punched this transhumanist in the ripcage and kicked her in the stomach',
      ],
      answers: [answer('ok', 'close'), attack()],
    },
  ]),
  restrictedAreaCheck: dialog((world, player) => {
    const answers: (Answer | Navigation)[] = [attack()]
    if (hasItem(world, player, 'idCard')) {
      answers.push(navigation('[show id card]', 1))
    }
    return [
      {
        text: ['Hey you, stop it right there. This is a restricted area.', 'Identify yourself'],
        answers,
      },
      {
        text: ['Ok, you can pass'],
        answers: [answer('ok', 'authorized')],
      },
    ]
  }),
  elevator: statelessDialog([
    {
      text: ['Elevator controls. Please select a level'],
      answers: [moveUp(), moveDown()],
    },
  ]),
}
export type DialogType = keyof typeof dialogDefinitions
export const dialogs: { [key in DialogType]: Dialog } = dialogDefinitions

export function addDialog(world: TlbWorld, entity: Entity, type: DialogType): Entity {
  const dialog = world
    .createEntity()
    .withComponent<TriggeredByComponent>('triggered-by', { entity })
    .withComponent<DialogComponent>('dialog', { type }).entity
  world.getComponent<TriggersComponent>(entity, 'triggers')!.entities.push(dialog)
  return dialog
}

const civilianRandomRemarks = [
  'we suspect that your sister controls the mutant rom chip and the frozen wand.',
  'I apoligize for this statement',
  'The CNT-FAI briefly considered joining the Comintern because they fetishized revolutionary upsurges and apparently knew nothing about politics or socialism',
  'Thinking about building a railgun',
  'Kinda want to hug, but like irl. for the propriocetive stimmy feels :P',
  'I am, at my core, a stubborn and frustrating human who only exists to be set in my ways.',
  "England are gonna win. And I'm gonna post so many garbage selfies tonight",
  'Love that my old man is spending his saturday swearing at remainers appearing on the news',
  "you think you do but you don't",
  'I reject nation states',
  'Woah I wish I could read',
  "What the hell is 'pajama'",
  'Always use protection',
  'Subsurface warfare is going to get lit',
  'Gracious wife, gracious life',
  'im about to nut these loads',
  'May biqueerplatonic people find success in their lives',
  'I hope hologender people are getting enough time to rest!',
  'Washing my chin',
  'I am still against full luxury automated communism. thats not what i aim for.',
  '#Javascript was a mistake.',
  'we had fun looking at all the medieval babies at the art gallery today',
  "Oh, me? I'm strictly chillin'",
  "It's a comfortable dusk in the city, and you are a comfortable salmon.",
  'adulthood is having your debit card declined, fashionably',
]

/*
00:00 Right. You
00:21 know, and
00:26 everything crisp
00:31 and we
00:36 just we
00:41 figured it
00:46 out. Black
00:59 Ops 2.
01:24 Then the just one
01:31 more input test by
01:38 it's also rifle team
01:45 could not using machine
01:52 in. Yeah,
02:05 yeah. Here's
02:31 one. Select.
03:03 As I said, This
03:09 is the fight fire
03:15 unwilling to tag just
03:21 put up this soon.
03:37 The app submitted fish relative quotes for the motions to
03:45 plant life conversion of toolbar come on his income considers
03:53 timberwolf does Metcalf burden for also 2, 1/2 or
04:00 and he deep sea teams and after the hunger plant.
04:07 Often the underplanted with fewer number? What
04:11 is the accentor after grammars perfecting it.
04:24 That's funny farm, he
04:31 bought all we would
04:38 miss him for us.
04:56 Then give me another disaster on boarding fast,
05:02 so the largest bank.
05:07 What is
05:27 life?
05:38 But I
05:53 will
06:01 not. Probably.
*/
