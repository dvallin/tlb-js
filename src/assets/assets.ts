import { Asset, Cover } from '../components/asset'
import { Vector } from '../spatial'
import { features, generators } from './features'
import { Feature } from '../components/feature'
import { DialogType } from './dialogs'

const s1x1 = new Vector([1, 1])
const s1x2 = new Vector([1, 2])
const s3x1 = new Vector([3, 1])
const s2x2 = new Vector([2, 2])
function asset(name: string, size: Vector, cover: Cover, dialog: DialogType | undefined, feature: (index: number) => Feature): Asset {
  return { name, size, cover, hasInventory: true, dialog, feature }
}

const assetsDefinition = {
  door: asset('metal door', s3x1, 'full', undefined, () => features['door']),
  locker: asset('a locker', s1x1, 'full', undefined, () => features['locker']),
  trash: asset('some trash', s1x1, 'none', undefined, () => features['trash']),
  loot: asset('a dead body', s1x1, 'none', undefined, () => features['loot']),
  table: asset('a table', s1x2, 'partial', undefined, () => features['table']),
  generator: asset('a generator', s2x2, 'full', undefined, generators['block']),
  elevator: asset('an elevator', s1x1, 'none', 'elevator', generators['elevator']),
}

export type AssetType = keyof typeof assetsDefinition
export const assets: { [key in AssetType]: Asset } = assetsDefinition
