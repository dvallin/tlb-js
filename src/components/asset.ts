import { Vector } from '../spatial'
import { TlbWorld } from '../tlb'
import { Entity } from '../ecs/entity'
import { WorldMap, WorldMapResource } from '../resources/world-map'
import { Shape } from '../geometry/shape'

import { FeatureComponent, createFeature, Feature, FeatureProvider } from './feature'
import { GroundComponent } from './ground'
import { PositionComponent } from './position'
import { Rectangle } from '../geometry/rectangle'
import { TriggersComponent } from './trigger'

import { AssetType, assets } from '../assets/assets'
import { FeatureType } from '../assets/features'
import { Direction } from '../spatial/direction'
import { InventoryComponent } from './items'
import { DialogType } from '../assets/dialogs'
import { DialogComponent } from './dialog'

export interface AssetComponent {
  type: AssetType
}

export interface Asset {
  name: string
  size: Vector
  hasInventory: boolean
  dialog: DialogType | undefined
  feature: (index: number) => Feature
}

export interface PlacedAsset {
  type: AssetType
  tiles: { position: Vector; feature: FeatureType }[]
}

export function shapeOfAsset(type: AssetType, position: Vector, direction: Direction): Shape {
  const asset = assets[type]
  return Rectangle.footprint(position, direction, asset.size)
}

export function createAsset(world: TlbWorld, level: number, position: Vector, direction: Direction, type: AssetType): Entity {
  const shape = shapeOfAsset(type, position, direction)
  return createAssetFromShape(world, level, shape, type)
}

export function createAssetFromShape(world: TlbWorld, level: number, shape: Shape, type: AssetType): Entity {
  const entity = placeAsset(world, type)
  placeAssetParts(world, type, entity, level, shape)
  return entity
}

function placeAsset(world: TlbWorld, type: AssetType): Entity {
  const asset = assets[type]
  const entity = world
    .createEntity()
    .withComponent<AssetComponent>('asset', { type })
    .withComponent<TriggersComponent>('triggers', { name: asset.name, type: 'asset', entities: [] }).entity
  if (asset.hasInventory) {
    world.editEntity(entity).withComponent<InventoryComponent>('inventory', { content: [] })
  }
  if (asset.dialog !== undefined) {
    world.editEntity(entity).withComponent<DialogComponent>('dialog', { type: asset.dialog })
  }
  return entity
}

function placeAssetParts(world: TlbWorld, type: AssetType, entity: Entity, level: number, shape: Shape): void {
  const map: WorldMap = world.getResource<WorldMapResource>('map')
  const asset = assets[type]
  shape.foreach((p, i) => {
    const tile = putTile(world, map, level, p, () => asset.feature(i))
    addTrigger(world, tile, entity)
  })
}

export function removeAsset(world: TlbWorld, map: WorldMap, entity: Entity): void {
  const triggers = world.getComponent<TriggersComponent>(entity, 'triggers')!
  triggers.entities.forEach(tile => removeTile(world, map, tile))
  world.deleteEntity(entity)
}

function removeTile(world: TlbWorld, map: WorldMap, entity: Entity): void {
  const ground = world.getComponent<GroundComponent>(entity, 'ground')!
  const position = world.getComponent<PositionComponent>(entity, 'position')!
  map.levels[position.level].removeTile(position.position)
  createFeature(world, position.level, position.position, ground.feature)
  world.deleteEntity(entity)
}

function putTile(world: TlbWorld, map: WorldMap, level: number, position: Vector, feature: FeatureProvider): Entity {
  const ground = removeGround(world, map, level, position)
  const tile = world
    .createEntity()
    .withComponent<PositionComponent>('position', { level, position })
    .withComponent<FeatureComponent>('feature', { feature })
    .withComponent<GroundComponent>('ground', { feature: ground }).entity
  map.levels[level].setTile(position, tile)
  return tile
}

function addTrigger(world: TlbWorld, tile: Entity, entity: Entity): void {
  world.editEntity(tile).withComponent('triggered-by', { entity })
  world.getComponent<TriggersComponent>(entity, 'triggers')!.entities.push(tile)
}

function removeGround(world: TlbWorld, map: WorldMap, level: number, position: Vector): FeatureProvider {
  const entity = map.levels[level].removeTile(position)
  if (entity === undefined) {
    throw new Error(`cannot build asset on missing ground at ${position.key}`)
  }
  const feature = world.getComponent<FeatureComponent>(entity, 'feature')
  if (feature === undefined || feature.feature().blocking) {
    throw new Error('cannot build asset on missing or blocking ground')
  }
  world.deleteEntity(entity)
  return feature.feature
}
