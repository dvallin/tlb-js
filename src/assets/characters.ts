import { BodyPart, CharacterStats, CharacterStatsComponent, createCharacterStatsComponent } from '../components/character-stats'
import { TlbWorld } from '../tlb'
import { features, FeatureType } from './features'
import { FeatureComponent } from '../components/feature'
import { FovComponent } from '../components/fov'
import { HasActionComponent } from '../components/action'
import { ItemComponent, InventoryComponent, EquipedItemsComponent } from '../components/items'
import { ActiveEffectsComponent } from '../components/effects'
import { CharacterCreator } from '../generative/complex-embedder'
import { Entity } from '../ecs/entity'
import { AiComponent } from '../components/ai'
import { ActionType } from './actions'
import { ItemType } from './items'
import { addDialog } from './dialogs'
import { TriggersComponent } from '../components/trigger'
import { SetSpace } from '../spatial/set-space'
import { WorldMapResource } from '../resources/world-map'

function humanoidBodyParts(core: number, head: number, limp: number): { [key: string]: BodyPart } {
  return {
    head: { type: 'head', itemAttachments: ['helmet'], health: head },
    leftArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health: limp },
    rightArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health: limp },
    torso: { type: 'torso', itemAttachments: ['armor'], health: core },
    leftLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health: limp },
    rightLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health: limp },
  }
}

const charactersStatsDefinition = {
  player: {
    bodyParts: humanoidBodyParts(10, 5, 5),
    strength: 5,
    aim: 10,
  },
  guard: {
    bodyParts: humanoidBodyParts(3, 1, 1),
    strength: 5,
    aim: 6,
  },
  civilian: {
    bodyParts: humanoidBodyParts(3, 1, 1),
    strength: 4,
    aim: 3,
  },
  eliteGuard: {
    bodyParts: humanoidBodyParts(5, 3, 3),
    strength: 8,
    aim: 6,
  },
  boss: {
    bodyParts: humanoidBodyParts(8, 5, 5),
    strength: 10,
    aim: 8,
  },
}
export type CharacterStatsType = keyof typeof charactersStatsDefinition
export const characterStats: { [key in CharacterStatsType]: CharacterStats } = charactersStatsDefinition

const characterCreatorsDefinition = {
  player: (world: TlbWorld) => createPlayer(world),
  eliteGuard: (world: TlbWorld) => createGuard(world, 'some elite guard', 'eliteGuard', 'eliteGuard'),
  guard: (world: TlbWorld) => createGuard(world, 'some guard', 'guard', 'guard'),
  civilian: (world: TlbWorld) => createCivilian(world, 'some guy', 'civilian', 'civilian'),
}

export const defaultActions: ActionType[] = ['move', 'hit', 'rush', 'endTurn']

export function createPlayer(world: TlbWorld): Entity {
  const player = createCharacter(world, 'player', 'player', defaultActions)
  const width = world.getResource<WorldMapResource>('map').width
  world
    .editEntity(player)
    .withComponent<{}>('player', {})
    .withComponent<FovComponent>('fov', { fov: new SetSpace(width) })
    .withComponent<{}>('viewport-focus', {})

  take(world, player, 'idCard')
  take(world, player, 'rifle')
  take(world, player, 'sniperRifle')
  take(world, player, 'bootsOfStriding')
  equip(world, player, 'nailGun', ['leftArm'])
  equip(world, player, 'leatherJacket', ['torso', 'leftArm', 'rightArm'])
  return player
}

export function createCivilian(world: TlbWorld, name: string, statsType: CharacterStatsType, featureType: FeatureType): Entity {
  const npc = createEmptyNpc(world, name, statsType, featureType, defaultActions)
  addDialog(world, npc, 'civilianDialog')
  return npc
}

export function createGuard(world: TlbWorld, name: string, statsType: CharacterStatsType, featureType: FeatureType): Entity {
  const npc = createEmptyNpc(world, name, statsType, featureType, defaultActions)
  world.editEntity(npc).withComponent<AiComponent>('ai', { type: 'rushing', state: 'idle', interest: undefined, distrust: 0 })
  equip(world, npc, 'rifle', ['leftArm'])
  addDialog(world, npc, 'guardRandomRemarks')
  return npc
}

export function take(world: TlbWorld, entity: Entity, type: ItemType): Entity {
  const item = world.createEntity().withComponent<ItemComponent>('item', { type }).entity
  world.getComponent<InventoryComponent>(entity, 'inventory')!.content.push(item)
  return item
}

export function equip(world: TlbWorld, entity: Entity, type: ItemType, bodyParts: string[]): void {
  const item = take(world, entity, type)
  world.getComponent<EquipedItemsComponent>(entity, 'equiped-items')!.equipment.push({
    entity: item,
    bodyParts,
  })
}

export function createEmptyNpc(
  world: TlbWorld,
  name: string,
  statsType: CharacterStatsType,
  featureType: FeatureType,
  actions: ActionType[]
): Entity {
  const character = createCharacter(world, statsType, featureType, actions)
  return world
    .editEntity(character)
    .withComponent('npc', {})
    .withComponent<TriggersComponent>('triggers', { entities: [], type: 'dialog', name }).entity
}

export function createCharacter(world: TlbWorld, statsType: CharacterStatsType, featureType: FeatureType, actions: ActionType[]): Entity {
  return world
    .createEntity()
    .withComponent<HasActionComponent>('has-action', { actions })
    .withComponent<InventoryComponent>('inventory', { content: [] })
    .withComponent<EquipedItemsComponent>('equiped-items', { equipment: [] })
    .withComponent<FeatureComponent>('feature', { feature: () => features[featureType] })
    .withComponent<CharacterStatsComponent>('character-stats', createCharacterStatsComponent(statsType))
    .withComponent<ActiveEffectsComponent>('active-effects', { effects: [] }).entity
}

export type CharacterCreatorType = keyof typeof characterCreatorsDefinition
export const characterCreators: { [key in CharacterCreatorType]: CharacterCreator } = characterCreatorsDefinition
