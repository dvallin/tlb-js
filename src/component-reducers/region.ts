import { TlbWorld } from '../tlb'
import { StructureComponent, RegionComponent } from '../components/region'
import { Entity } from '../ecs/entity'
import { PositionComponent } from '../components/position'
import { WorldMap, WorldMapResource } from '../resources/world-map'

export function isAuthorized(world: TlbWorld, position: PositionComponent, entity: Entity): boolean {
  const map: WorldMap = world.getResource<WorldMapResource>('map')
  const structure = map.levels[position.level].getStructure(position.position)!
  const room: StructureComponent = world.getComponent<StructureComponent>(structure, 'structure')!
  const region: RegionComponent = world.getComponent<RegionComponent>(room.region, 'region')!
  return region.authorized.has(entity)
}

export function authorize(world: TlbWorld, position: PositionComponent, entity: Entity): void {
  const map: WorldMap = world.getResource<WorldMapResource>('map')
  const structure = map.levels[position.level].getStructure(position.position)!
  const room: StructureComponent = world.getComponent<StructureComponent>(structure, 'structure')!
  const region: RegionComponent = world.getComponent<RegionComponent>(room.region, 'region')!
  region.authorized.add(entity)
}
