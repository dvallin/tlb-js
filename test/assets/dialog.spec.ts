import { TlbWorld, registerComponents } from '../../src/tlb'
import { Entity } from '../../src/ecs/entity'
import { WorldMap, WorldMapResource } from '../../src/resources/world-map'
import { Uniform } from '../../src/random/distributions'
import { characterCreators } from '../../src/assets/characters'
import { Vector } from '../../src/spatial'
import { placeCharacter } from '../../src/component-reducers/place-character'
import { World } from '../../src/ecs/world'
import { createFeatureFromType } from '../../src/components/feature'
import { StructureComponent, RegionComponent } from '../../src/components/region'
import { createAssetFromShape } from '../../src/components/asset'
import { Rectangle } from '../../src/geometry/rectangle'
import { Random } from '../../src/random'
import { dialogs } from '../../src/assets/dialogs'

describe('dialog', () => {
  describe('terminal', () => {
    let world: TlbWorld
    let player: Entity
    let terminal: Entity

    let map: WorldMap
    let region: Entity
    beforeEach(() => {
      world = new World()
      registerComponents(world)

      world.registerResource(new WorldMapResource(4))
      map = world.getResource<WorldMapResource>('map')

      const boundary = map.levels[0].boundary
      region = world.createEntity().withComponent<RegionComponent>('region', {
        type: 'red',
        shape: boundary,
        level: 0,
        authorized: new Set(),
        exits: [],
      }).entity
      const room = world
        .createEntity()
        .withComponent<StructureComponent>('structure', { region, kind: 'room', shape: boundary, connections: [] }).entity
      boundary.foreach(p => {
        createFeatureFromType(world, 0, p, 'corridor')
        map.levels[0].setStructure(p, room)
      })

      player = characterCreators.player(world)
      placeCharacter(world, player, 0, new Vector([0, 0]))

      const random = new Random(new Uniform('terminal'))
      terminal = createAssetFromShape(world, random, 0, new Rectangle(0, 1, 1, 1), 'terminal')
      createAssetFromShape(world, random, 0, new Rectangle(1, 0, 1, 1), 'door')
      createAssetFromShape(world, random, 0, new Rectangle(1, 1, 1, 1), 'door')
    })

    it('shows a main menu', () => {
      const steps = dialogs.terminal.steps(world, player, terminal)

      expect(steps[0]).toMatchSnapshot()
    })

    it('shows doors', () => {
      const steps = dialogs.terminal.steps(world, player, terminal)

      const doorControls = steps.find(s => s.text[0] === 'door controls')!
      expect(doorControls).toMatchSnapshot()
    })

    it('shows lore', () => {
      const steps = dialogs.terminal.steps(world, player, terminal)

      const lore = steps.find(s => s.text[0] === 'lore')!
      expect(lore).toMatchSnapshot()
    })
  })
})
