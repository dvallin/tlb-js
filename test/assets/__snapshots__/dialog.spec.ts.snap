// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dialog terminal shows a main menu 1`] = `
Object {
  "answers": Array [
    Object {
      "dialog": "startLore1",
      "text": "lore.txt",
      "type": "next_dialog",
    },
    Object {
      "text": "door controls",
      "type": 1,
    },
    Object {
      "target": undefined,
      "text": "exit",
      "type": "close",
    },
  ],
  "text": Array [
    "terminal",
  ],
}
`;

exports[`dialog terminal shows doors 1`] = `
Object {
  "answers": Array [
    Object {
      "target": 26,
      "text": "a door to the right",
      "type": "trigger",
    },
    Object {
      "target": 27,
      "text": "a door to the right,down",
      "type": "trigger",
    },
    Object {
      "text": "back",
      "type": 0,
    },
  ],
  "text": Array [
    "door controls",
  ],
}
`;

exports[`dialog terminal shows lore 1`] = `undefined`;
